File: backend/src/scripts/generateLocalEmbeddings.ts
// generateLocalEmbeddings.ts
import * as fs from 'fs';
import * as path from 'path';
import { GoogleGenerativeAI } from '@google/generative-ai';
import dotenv from 'dotenv';

// Load environment variables
dotenv.config({ path: path.resolve(__dirname, '../../.env') });

// Configuration
const GEMINI_API_KEY = process.env.GOOGLE_API_KEY;
const SOURCE_DIR = path.resolve(__dirname, '../../../FF-BASE');
const OUTPUT_FILE = path.resolve(__dirname, '../../../knowledge_base/embeddings.json');

if (!GEMINI_API_KEY) {
  console.error('GOOGLE_API_KEY is not set in environment variables');
  process.exit(1);
}

const genAI = new GoogleGenerativeAI(GEMINI_API_KEY);
const embeddingModel = genAI.getGenerativeModel({ model: 'embedding-001' });

interface EmbeddingData {
  file_path: string;
  content: string;
  embedding: number[];
}

/**
 * Get all .md files from directory recursively
 */
function getAllMdFiles(dir: string, fileList: string[] = []): string[] {
  const files = fs.readdirSync(dir);
  
  files.forEach(file => {
    const filePath = path.join(dir, file);
    const stat = fs.statSync(filePath);
    
    // Skip hidden directories
    if (stat.isDirectory() && !file.startsWith('.')) {
      getAllMdFiles(filePath, fileList);
    } else if (stat.isFile() && file.endsWith('.md')) {
      fileList.push(filePath);
    }
  });
  
  return fileList;
}

/**
 * Read file content
 */
function readFileContent(filePath: string): string {
  try {
    return fs.readFileSync(filePath, 'utf-8');
  } catch (error) {
    console.error(`Error reading file ${filePath}:`, error);
    return '';
  }
}

/**
 * Generate embedding for text
 */
async function generateEmbedding(text: string): Promise<number[]> {
  try {
    // Limit text size for API
    if (text.length > 10000) {
      text = text.substring(0, 10000);
    }
    
    const result = await embeddingModel.embedContent(text);
    return result.embedding.values;
  } catch (error) {
    console.error('Error generating embedding:', error);
    return [];
  }
}

/**
 * Generate embeddings for all .md files in directory
 */
async function generateEmbeddingsForDirectory(): Promise<void> {
  console.log(`Searching for .md files in directory: ${SOURCE_DIR}`);
  
  // Get all .md files
  const files = getAllMdFiles(SOURCE_DIR);
  console.log(`Found ${files.length} .md files`);
  
  const embeddingsData: EmbeddingData[] = [];
  
  // Process each file
  for (let i = 0; i < files.length; i++) {
    const filePath = files[i];
    const relativePath = path.relative(SOURCE_DIR, filePath);
    
    console.log(`Processing file ${i + 1}/${files.length}: ${relativePath}`);
    
    // Read file content
    const content = readFileContent(filePath);
    if (!content) {
      continue;
    }
    
    // Generate embedding
    const embedding = await generateEmbedding(content);
    if (embedding.length > 0) {
      embeddingsData.push({
        file_path: relativePath,
        content: content,
        embedding: embedding
      });
    } else {
      console.log(`Failed to generate embedding for ${relativePath}`);
    }
    
    // Show progress every 10 files
    if ((i + 1) % 10 === 0) {
      console.log(`Processed ${i + 1} files...`);
    }
  }
  
  // Save embeddings to file
  console.log(`Saving ${embeddingsData.length} embeddings to ${OUTPUT_FILE}`);
  
  // Ensure directory exists
  const dir = path.dirname(OUTPUT_FILE);
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }
  
  fs.writeFileSync(OUTPUT_FILE, JSON.stringify(embeddingsData, null, 2), 'utf-8');
  console.log('Embeddings saved successfully!');
}

// Run the script
async function main() {
  try {
    console.log('Starting embeddings generation for local files...');
    await generateEmbeddingsForDirectory();
    console.log('Process completed successfully!');
  } catch (error) {
    console.error('Error:', error);
    process.exit(1);
  }
}

main();

File: python/generate_local_embeddings.py
import os
import json
import glob
from dotenv import load_dotenv
import google.generativeai as genai

# Загрузка переменных окружения из .env файла
# Load from root .env file first
env_file = Path("../.env")
if env_file.exists():
    load_dotenv(env_file)
    print("Environment variables loaded from root .env")
else:
    # Fallback to backend/.env
    backend_env_file = Path("../backend/.env")
    if backend_env_file.exists():
        load_dotenv(backend_env_file)
        print("Environment variables loaded from backend/.env")

# Получение токенов и параметров из переменных окружения
GEMINI_API_KEY = os.getenv("GOOGLE_API_KEY")

# Настройка Google Gemini
if GEMINI_API_KEY:
    genai.configure(api_key=GEMINI_API_KEY)
    embedding_model = "models/embedding-001"
else:
    print("Error: GEMINI_API_KEY not found")
    exit(1)

# Путь к файлу для хранения эмбеддингов
EMBEDDINGS_FILE = "knowledge_base/embeddings.json"

def get_local_files(root_dir):
    """Получение списка всех .md файлов из локальной директории"""
    print(f"Searching for .md files in directory: {root_dir}")
    
    md_files = []
    for root, dirs, files in os.walk(root_dir):
        # Пропускаем скрытые директории
        dirs[:] = [d for d in dirs if not d.startswith('.')]
        for file in files:
            if file.endswith('.md'):
                full_path = os.path.join(root, file)
                relative_path = os.path.relpath(full_path, root_dir)
                md_files.append({
                    "path": relative_path,
                    "full_path": full_path
                })
    
    print(f"Found {len(md_files)} .md files")
    return md_files

def get_file_content(file_path):
    """Получение содержимого файла"""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            return f.read()
    except Exception as e:
        print(f"Error reading file {file_path}: {e}")
        return ""

def generate_embedding(text):
    """Генерация эмбеддинга для текста с помощью Google Gemini"""
    if not GEMINI_API_KEY:
        print("Error: GEMINI_API_KEY not found")
        return None
    
    try:
        # Ограничиваем размер текста для API
        if len(text) > 10000:
            text = text[:10000]
            
        result = genai.embed_content(
            model=embedding_model,
            content=text,
            task_type="retrieval_document"
        )
        return result["embedding"]
    except Exception as e:
        print(f"Error generating embedding: {e}")
        return None

def create_embeddings_for_local_files(source_dir, output_file):
    """Создание эмбеддингов для всех .md файлов в локальной директории"""
    # Получение списка файлов
    files = get_local_files(source_dir)
    if not files:
        print("No .md files found")
        return
    
    embeddings_data = []
    
    # Обработка каждого файла
    for i, file_info in enumerate(files):
        file_path = file_info["path"]
        full_path = file_info["full_path"]
        
        print(f"Processing file {i+1}/{len(files)}: {file_path}")
        
        # Получение содержимого файла
        content = get_file_content(full_path)
        if not content:
            continue
        
        # Генерация эмбеддинга для содержимого файла
        embedding = generate_embedding(content)
        if embedding:
            # Сохранение данных об эмбеддинге
            embeddings_data.append({
                "file_path": file_path,
                "content": content,
                "embedding": embedding
            })
        else:
            print(f"Failed to generate embedding for {file_path}")
    
    # Сохранение эмбеддингов в файл
    print(f"Saving {len(embeddings_data)} embeddings to {output_file}")
    os.makedirs(os.path.dirname(output_file), exist_ok=True)
    
    with open(output_file, "w", encoding="utf-8") as f:
        json.dump(embeddings_data, f, ensure_ascii=False, indent=2)
    
    print("Embeddings saved successfully!")
    return len(embeddings_data)

if __name__ == "__main__":
    source_directory = os.getenv("FF_BASE_DIR", "/Users/eugene/Library/CloudStorage/GoogleDrive-ekirshin@gmail.com/Мой диск/OBSIDIAN/FF-BASE")
    output_file = "knowledge_base/embeddings.json"
    
    print("Starting embeddings generation for local files...")
    result = create_embeddings_for_local_files(source_directory, output_file)
    print(f"Process completed. Generated embeddings for {result} files.")